### PR TITLE
Add data tiers (hot, warm, cold, frozen) as custom node roles

### PR DESCRIPTION
**Pull request by dakrone**
This commit adds the `data_hot`, `data_warm`, `data_cold`, and `data_frozen` node roles to the
x-pack plugin. These roles are intended to be the base for the formalization of data tiers in
Elasticsearch.

These roles all act as data nodes (meaning shards can be allocated to them). Nodes with the existing
`data` role acts as though they have all of the roles configured (it is a hot, warm, cold, and
frozen node).

This also includes a custom `AllocationDecider` that allows the user to configure the following
settings on a cluster level:
- `cluster.routing.allocation.require._tier`
- `cluster.routing.allocation.include._tier`
- `cluster.routing.allocation.exclude._tier`

And in index settings:
- `index.routing.allocation.require._tier`
- `index.routing.allocation.include._tier`
- `index.routing.allocation.exclude._tier`

Relates to #60848

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+    @Override
+    public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return shouldFilter(shardRouting, node.node(), allocation);
+    }
+
+    @Override
+    public Decision shouldAutoExpandToNode(IndexMetadata indexMetadata, DiscoveryNode node, RoutingAllocation allocation) {
+        Decision decision = shouldClusterFilter(node, allocation);
+        if (decision != null) return decision;
```
Comment (andreidan) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:95]: nit: maybe personal preference but I find if statements without { } error-prone ( I seem to remember apple was bitten by this https://www.imperialviolet.org/2014/02/22/applebug.html )
There are a few if statements like this in this class, I'd rather have all of them use curly braces

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDeciderTests.java [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+    static {
+        Set<Setting<?>> allSettings = new HashSet<>(ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        allSettings.add(DataTierAllocationDecider.CLUSTER_ROUTING_REQUIRE_SETTING);
+        allSettings.add(DataTierAllocationDecider.CLUSTER_ROUTING_INCLUDE_SETTING);
+        allSettings.add(DataTierAllocationDecider.CLUSTER_ROUTING_EXCLUDE_SETTING);
+        ALL_SETTINGS = allSettings;
+    }
+
+    public void testClusterRequires() {
```
Comment (andreidan) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDeciderTests.java:73]: shall we add a test that has both cluster and index routings? (maybe one where allocation is successful and one where the index routing matches but the cluster one doesn't to test the cluster settings precedence?

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 3]
```diff
@@ -72,8 +73,13 @@ public static boolean isMasterNode(final Settings settings) {
         return hasRole(settings, DiscoveryNodeRole.MASTER_ROLE);
     }
 
+    /**
+     * Note, due to the way that plugins may not be available when settings are being initialized,
+     * do not use this from a static {@link Setting} default value function, as it may not contain
+     * all roles at that point.
```
Comment (andreidan) [inline server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:79]: should we mention the `hasRole || DataTier.isExplicitDataTier` alternative in the doc here?

**Review by andreidan — COMMENTED** [review]
Thanks for working on this Lee. Left some rather minor questions and suggestion.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java (hunk 1 above)
Comment (dakrone) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java:95]: Yes I agree, this was copied from `FilterAllocationDecider` which is why, I'll change these to real `if` statements

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java (hunk 3 above)
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:79]: `DataTier` is an x-pack concept, so I didn't want to mention it here

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java [hunk 4]
```diff
@@ -72,8 +73,13 @@ public static boolean isMasterNode(final Settings settings) {
         return hasRole(settings, DiscoveryNodeRole.MASTER_ROLE);
     }
 
+    /**
+     * Note, due to the way that plugins may not be available when settings are being initialized,
```
[the remark this answers is not in the document]
Comment (dakrone) [reply server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java:77]: I'll reword this a bit, thanks for the suggestion

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDeciderTests.java (hunk 2 above)
Comment (dakrone) [reply x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDeciderTests.java:73]: Sure! I added this

File: server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (andreidan) [reply server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNode.java]: Thanks for updating this 👍

**Review by andreidan — APPROVED** [review]
LGTM, thanks for iterating on this Lee

**Comment by dakrone** [comment]
This commit adds the `data_hot`, `data_warm`, `data_cold`, and `data_frozen` node roles to the
x-pack plugin. These roles are intended to be the base for the formalization of data tiers in
Elasticsearch.

These roles all act as data nodes (meaning shards can be allocated to them). Nodes with the existing
`data` role acts as though they have all of the roles configured (it is a hot, warm, cold, and
frozen node).

This also includes a custom `AllocationDecider` that allows the user to configure the following
settings on a cluster level:
- `cluster.routing.allocation.require._tier`
- `cluster.routing.allocation.include._tier`
- `cluster.routing.allocation.exclude._tier`

And in index settings:
- `index.routing.allocation.require._tier`
- `index.routing.allocation.include._tier`
- `index.routing.allocation.exclude._tier`

Relates to #60848

**Pull request merged**